### PR TITLE
[DataStorage] API for Multiple Reading based on start and end timestamp

### LINKED ISSUE #286: [Data Storage] Provide an API in Home Edge to store device generated data
Existing Problem:
Currently if any developer (using Rpi/Sensor) want to store device generated data using Home Edge, it is not possible to do. The developer has to again use EdgeX API/SDK to store the data. This makes the developer to know the working of both Home Edge and EdgeX. Home Edge should provide seamless discovery, orchestration and data sharing mechanism. For developer Home Edge should be the only point of consideration and need not bother on other solutions used internally. 

Solution:
1. Home Edge can provide an API ( POST) which can be used by developers to store the device data. The API should provide methods to store int, float, string, binary(image) data. This API can internally leverage on EdgeX to store the data.
2. We should also have an APi to get the stored values of a particular device .

### LINKED ISSUE #319: [DataStorage] Need APIs for readings
- [x] **POST/GET** API route for a reading without deviceNameKey : clients.ApiBase + "/resource/{" + resourceNameKey + "}"

    - Set [deviceName](https://github.com/lf-edge/edge-home-orchestration-go/blob/ca7ca4c92df7362bd09d0c83dba22fc21ae3863c/internal/controller/storagemgr/storagedriver/storagehandler.go#L126)  from the yaml configuration or a kind of function like [`GetDeviceID`](https://github.com/lf-edge/edge-home-orchestration-go/blob/61c218e3c0d24c8b457876245b78f5f0abcc380b/internal/controller/discoverymgr/discovery.go#L75).
    - Use deviceName for a [reading](https://github.com/lf-edge/edge-home-orchestration-go/blob/ca7ca4c92df7362bd09d0c83dba22fc21ae3863c/internal/controller/storagemgr/storagedriver/storagehandler.go#L112) from edgex-data

- [ ] GET API for readings in terms of time
    - Related edgex API(vHanoi) : [/v1/reading/{start}/{end}/{limit}](https://app.swaggerhub.com/apis-docs/EdgeXFoundry1/core-data/1.2.1#/default/get_v1_reading__start___end___limit_)
 
- [x] GET API for readings in terms of resource type
    - Related edgex API(vHanoi) :  [/v1/reading/name/{name}/device/{device}/{limit}](https://app.swaggerhub.com/apis-docs/EdgeXFoundry1/core-data/1.2.1#/default/get_v1_reading_name__name__device__device___limit_)

### PR DESCRIPTION
**Pull request by nitu-s-gupta**
Added an API for fetching multiple readings from a device at once based on start and end timestamp

Usage Example
**curl --location --request GET http://localhost:49986/api/v1/start/{start}/end/{end}/{ReadingCount}**

Signed-off-by: Nitu Gupta <nitu.gupta@samsung.com>

# Description
This is a GET API used to get multiple stored readings of any type stored by a device based on start and end timestamp

Fixes #319 

## Type of change
- [X] New feature (non-breaking change which adds functionality)

# How Has This Been Tested?
To test we tried to retrieve 2 and more instances of stored data using

curl --location --request GET 'http://localhost:49986/api/v1/start/1629733474672/end/1629782168113/10'

The output was received as desired
[
    {
        "id": "26ce8920-b691-4209-81be-98b5d7291947",
        "created": 1629733474672,
        "origin": 1629733474671902964,
        "device": "datastorage",
        "name": "string",
        "value": "raw data2",
        "valueType": "String"
    },
    {
        "id": "c76a7c20-66d6-4bc9-bbd5-71bd6bf231de",
        "created": 1629733481630,
        "origin": 1629733481629912818,
        "device": "datastorage",
        "name": "string",
        "value": "raw data3",
        "valueType": "String"
    },
    {
        "id": "2ad7e8fd-a542-463f-93c1-f90711aa9f72",
        "created": 1629733721156,
        "origin": 1629733721155217982,
        "device": "datastorage",
        "name": "string",
        "value": "dfjbvfkjbgjkhnvgjkehgjkehgjktehnjkg",
        "valueType": "String"
    },
    {
        "id": "8df656c2-343e-45d1-812f-caf304195be9",
        "created": 1629782147086,
        "origin": 1629782147085206683,
        "device": "datastorage",
        "name": "string",
        "value": "dfjbvfkjbgjkhnvgjkehgjkehgjktehnjkg",
        "valueType": "String"
    },
    {
        "id": "4ad617bc-2906-4faa-be05-2a1b5199956a",
        "created": 1629782168113,
        "origin": 1629782168112423932,
        "device": "datastorage",
        "name": "string",
        "value": "test string",
        "valueType": "String"
    }
]

**Test Configuration**:
* Firmware version: Ubuntu 20.04
* Hardware: x86-64
* Edge Orchestration Release: 1.0.0

 # Checklist:

- [X] My code follows the style guidelines of this project
- [X] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [X] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [X] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: internal/controller/storagemgr/storagedriver/storagehandler.go [hunk 1]
```diff
@@ -107,46 +113,52 @@ func (handler StorageHandler) addContext(next func(http.ResponseWriter, *http.Re
...
-			return
-		}
-	}
+	start := vars[startKey]
+	end := vars[endKey]
 	resourceName := vars[resourceNameKey]
 	readingCount := vars[numberOfReadings]
-
+	readingAPI := ""
+	log.Println(vars)
```
Comment (t25kim) [inline internal/controller/storagemgr/storagedriver/storagehandler.go:121]: This seems unnecessary.

File: internal/controller/storagemgr/storagedriver/storagehandler.go [hunk 2]
```diff
@@ -107,46 +113,52 @@ func (handler StorageHandler) addContext(next func(http.ResponseWriter, *http.Re
...
 	if readingCount == "" {
 		readingCount = "1"
 	}
+	var err error
+	if start != "" && end != "" {
+		log.Debug(fmt.Sprintf("Received Get for start =%s end=%s", start, end))
```
Comment (t25kim) [inline internal/controller/storagemgr/storagedriver/storagehandler.go:127]:
```suggestion
		log.Debug(fmt.Sprintf("Received Get for start=%s end=%s", start, end))
```

File: internal/controller/storagemgr/storagedriver/storagehandler.go [hunk 3]
```diff
@@ -46,11 +46,14 @@ const (
 	handlerContextKey ctxKey = "StorageHandler"
 	configPath               = "res/configuration.toml"
 	numberOfReadings         = "readingCount"
-
-	apiResourceRoute           = clients.ApiBase + "/device/{" + deviceNameKey + "}/resource/{" + resourceNameKey + "}"
-	apiResourceRouteMultiple   = clients.ApiBase + "/device/{" + deviceNameKey + "}/resource/{" + resourceNameKey + "}/{" + numberOfReadings + "}"
-	apiWithoutResRoute         = clients.ApiBase + "/resource/{" + resourceNameKey + "}"
-	apiWithoutResRouteMultiple = clients.ApiBase + "/resource/{" + resourceNameKey + "}/{" + numberOfReadings + "}"
+	startKey                 = "start"
+	endKey                   = "end"
```
Comment (t25kim) [inline internal/controller/storagemgr/storagedriver/storagehandler.go:50]: How about keeping keys together? I'd like to suggest to move start and end keys after resourceNameKey.

**Review by t25kim — COMMENTED** [review]
Thanks for contribution!
Please take a look at my comments.

File: internal/controller/storagemgr/storagedriver/storagehandler.go (hunk 3 above)
Comment (nitu-s-gupta) [reply internal/controller/storagemgr/storagedriver/storagehandler.go]: Done!

File: internal/controller/storagemgr/storagedriver/storagehandler.go (hunk 2 above)
Comment (nitu-s-gupta) [reply internal/controller/storagemgr/storagedriver/storagehandler.go]: Done

File: internal/controller/storagemgr/storagedriver/storagehandler.go (hunk 1 above)
Comment (nitu-s-gupta) [reply internal/controller/storagemgr/storagedriver/storagehandler.go]: Done!

**Review by t25kim — APPROVED** [review]
LGTM Thanks!

**Review by MoonkiHong — APPROVED** [review]
+1

**Review by tdrozdovsky — APPROVED** [review]
LGTM

**Review by suresh-lc — APPROVED** [review]
LGTM, thanks for your contributions

**Pull request merged**